sycl/source/detail/queue_impl.cpp

```diff
@@ -97,9 +97,20 @@ event queue_impl::mem_advise(const shared_ptr_class<detail::queue_impl> &Self,
 }
 
 void queue_impl::addEvent(const event &Event) {
-  std::weak_ptr<event_impl> EventWeakPtr{getSyclObjImpl(Event)};
-  std::lock_guard<mutex_class> Lock(MMutex);
-  MEvents.push_back(std::move(EventWeakPtr));
+  // if the command behind the event has no memory dependencies,
+  // we need to track the event with the USMEvents, or it won't be properly
+  // released.
+  EventImplPtr Eimpl = getSyclObjImpl(Event);
+  Command *Cmd = (Command *)(Eimpl->getCommand());
+  if (Cmd && Cmd->MDeps.size() == 0) {
+    addUSMEvent(Event);
+    Eimpl->setCommand(nullptr); // decouple and free the command
+    delete Cmd;
```

Comment:
Can this be done inside the scheduler? Just inside addCG?

Reply:
No.  That causes the resources to be released because the shared pointer will have no owner.  The call were are in, `addEvent` is called later, after `addCG` has completed. We need to first put the Event in the USMEvents vector and then delete the command.

Reply:
How so? `Scheduler::addCG` returns a shared pointer to the event associated with the created command, that shared pointer is then returned to `handler::finalize()` and then to `queue_impl::submit_impl(...)`. Why can't we kill off the command in `Scheduler::addCG` while keeping the `event` alive via that shared pointer?

Reply:
you are right, it's the same call stack. I was remembering from when I was originally trying to delete in addCG.  I can make that change.